plugins/inputs/azure_monitor/azure_monitor.go

```diff
@@ -0,0 +1,856 @@
+		fullResourceID := strings.Split(value.(map[string]interface{})["id"].(string), "/")
+		resourceID := strings.Join(fullResourceID[3:], "/")
+		resourceType := value.(map[string]interface{})["type"].(string)
+		resourceIndex := target.getResourceWithResourceTypeIndex(resourceType)
+
+		if resourceIndex == -1 {
+			continue
+		}
+
+		resourceTarget := NewResourceTarget(resourceID, target.Resources[resourceIndex].Metrics, target.Resources[resourceIndex].Aggregation)
+
+		am.ResourceTargets = append(am.ResourceTargets, resourceTarget)
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) getResourceTargetsMetrics() error {
+	var waitGroup sync.WaitGroup
+
+	errChan := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer cancel()
+
+	for index, target := range am.ResourceTargets {
+		if len(target.Metrics) > 0 {
+			continue
+		}
+
+		am.Log.Info("Getting metrics for target #", index+1)
+		waitGroup.Add(1)
+
+		go func(target *ResourceTarget) {
+			defer waitGroup.Done()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			err := am.getResourceTargetMetrics(target)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+		}(target)
+	}
+
+	waitGroup.Wait()
+
+	select {
+	case err := <-errChan:
+		return err
+	default:
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) getResourceTargetMetrics(target *ResourceTarget) error {
+	apiURL := am.buildMetricDefinitionsAPIURL(target)
+	body, err := am.getTargetResponseBody(apiURL)
+
+	if err != nil {
+		return err
+	}
+
+	err = target.setResourceTargetMetrics(body)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) checkTargetsMaximumMetrics() {
+	for _, target := range am.ResourceTargets {
+		if len(target.Metrics) <= maxMetricsPerRequest {
+			continue
+		}
+
+		for start := maxMetricsPerRequest; start < len(target.Metrics); start += maxMetricsPerRequest {
+			end := start + maxMetricsPerRequest
+
+			if end > len(target.Metrics) {
+				end = len(target.Metrics)
+			}
+
+			newTargetMetrics := target.Metrics[start:end]
+
+			var newTargetAggregation []string
+
+			for _, aggregation := range target.Aggregation {
+				newTargetAggregation = append(newTargetAggregation, aggregation)
+			}
+
+			newTarget := NewResourceTarget(target.ResourceID, newTargetMetrics, newTargetAggregation)
+
+			am.ResourceTargets = append(am.ResourceTargets, newTarget)
+		}
+
+		target.Metrics = target.Metrics[:maxMetricsPerRequest]
+	}
+}
+
+func (am *AzureMonitor) setResourceTargetsAggregation() {
+	for _, target := range am.ResourceTargets {
+		if len(target.Aggregation) > 0 {
+			continue
+		}
+
+		target.Aggregation = append(target.Aggregation, "Total", "Count", "Average", "Minimum", "Maximum")
+	}
+}
+
+func (am *AzureMonitor) Gather(acc telegraf.Accumulator) error {
+	err := am.refreshAccessToken()
+
+	if err != nil {
+		return err
+	}
```

Comment:
Shouldn't this be a goroutine started in e.g. `Init()`?

Reply:
No, each access token is valid for amount of time, and we need to check if we need to create a new one. Without it we won't be able to collect metrics from Azure

Reply:
Understood. Can you please add this information as a small comment in the code?